File: src/main/java/dev/esophose/discordbot/command/commands/misc/SudoCommand.java
package dev.esophose.discordbot.command.commands.misc;

import dev.esophose.discordbot.Sparky;
import dev.esophose.discordbot.command.DiscordCommand;
import dev.esophose.discordbot.command.DiscordCommandMessage;
import dev.esophose.discordbot.manager.CommandManager;
import dev.esophose.discordbot.manager.GuildSettingsManager;
import dev.esophose.discordbot.webhook.WebhookUtils;
import discord4j.core.object.entity.Member;
import discord4j.core.object.util.Permission;
import discord4j.core.object.util.PermissionSet;
import java.util.Collections;
import java.util.List;
import reactor.core.publisher.Mono;

public class SudoCommand extends DiscordCommand {

    public SudoCommand(Sparky bot) {
        super(bot);
    }

    public void execute(DiscordCommandMessage message, Member member, String command) {
        CommandManager commandManager = this.bot.getManager(CommandManager.class);

        String commandPrefix = this.bot.getManager(GuildSettingsManager.class).getGuildSettings(message.getGuildId()).getCommandPrefix();
        member.getAvatar().flatMap(avatar -> WebhookUtils.createAndExecuteWebhook(message.getChannel(), member.getDisplayName(), avatar, spec -> {
            spec.setContent(command);
        })).thenEmpty(Mono.fromRunnable(() -> {
            if (command.startsWith(commandPrefix))
                commandManager.executeCommand(message.getGuild(), message.getChannel(), message.getChannelId(), message.getMessageId(), member, command, commandPrefix);
        })).subscribe();
    }

    @Override
    public String getName() {
        return "sudo";
    }

    @Override
    public List<String> getAliases() {
        return Collections.singletonList("force");
    }

    @Override
    public String getDescription() {
        return "Performs a command as another member";
    }

    @Override
    public PermissionSet getRequiredBotPermissions() {
        return PermissionSet.of(Permission.SEND_MESSAGES, Permission.MANAGE_WEBHOOKS);
    }

    @Override
    public Permission getDefaultRequiredMemberPermission() {
        return Permission.ADMINISTRATOR;
    }

}

File: src/main/kotlin/dev/esophose/discordbot/command/commands/misc/SudoCommand.kt
package dev.esophose.discordbot.command.commands.misc

import dev.esophose.discordbot.Sparky
import dev.esophose.discordbot.command.DiscordCommand
import dev.esophose.discordbot.command.DiscordCommandMessage
import dev.esophose.discordbot.manager.CommandManager
import dev.esophose.discordbot.manager.GuildSettingsManager
import discord4j.core.`object`.entity.Member
import discord4j.core.`object`.entity.channel.TextChannel
import discord4j.rest.util.Permission
import discord4j.rest.util.PermissionSet
import reactor.core.publisher.Mono

class SudoCommand : DiscordCommand() {

    override val name: String
        get() = "sudo"

    override val aliases: List<String>
        get() = listOf("force")

    override val description: String
        get() = "Performs a command as another member"

    override val requiredBotPermissions: PermissionSet
        get() = PermissionSet.of(Permission.SEND_MESSAGES, Permission.MANAGE_WEBHOOKS)

    override val defaultRequiredMemberPermission: Permission
        get() = Permission.ADMINISTRATOR

    fun execute(message: DiscordCommandMessage, member: Member, command: String) {
        message.delete().subscribe()

        val commandManager = Sparky.getManager(CommandManager::class)
        val commandPrefix = Sparky.getManager(GuildSettingsManager::class).getGuildSettings(message.guildId).commandPrefix
        member.avatar.flatMap { avatar ->
            message.channel.cast(TextChannel::class.java).flatMap { channel ->
                channel.createWebhook { spec ->
                    spec.setName(member.displayName)
                    spec.setAvatar(avatar)
                }.flatMap { webhook ->
                    webhook.execute { spec ->
                        spec.setContent(command)
                    }.then(webhook.delete())
                }
            }
        }.thenEmpty(Mono.fromRunnable {
            if (command.startsWith(commandPrefix))
                commandManager.executeCommand(message.guild, message.channel, message.channelId, message.messageId, member, command, commandPrefix)
        }).subscribe()
    }

}
